// unzipper-esm/dist/Open/directory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = centralDirectory;
var _path = _interopRequireDefault(require("path"));
var _fs = _interopRequireDefault(require("fs"));
var _unzip = _interopRequireDefault(require("./unzip.js"));
var _mapPromises = _interopRequireDefault(require("../mapPromises.js"));
var _PullStream = _interopRequireDefault(require("../PullStream.js"));
var _BufferStream = _interopRequireDefault(require("../BufferStream.js"));
var _parseExtraField = _interopRequireDefault(require("../parseExtraField.js"));
var _parseDateTime = _interopRequireDefault(require("../parseDateTime.js"));
var _parseBuffer = _interopRequireDefault(require("../parseBuffer.js"));
var _ensureDir = _interopRequireDefault(require("../ensureDir.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; } // 'node:path'
// 'node:fs'
var signature = Buffer.alloc(4);
signature.writeUInt32LE(0x06054b50, 0);
function getCrxHeader(source) {
  var sourceStream = source.stream(0).pipe(new _PullStream["default"]());
  return sourceStream.pull(4).then(function (data) {
    var signature = data.readUInt32LE(0);
    if (signature === 0x34327243) {
      var crxHeader;
      return sourceStream.pull(12).then(function (data) {
        crxHeader = (0, _parseBuffer["default"])(data, [['version', 4], ['pubKeyLength', 4], ['signatureLength', 4]]);
      }).then(function () {
        return sourceStream.pull(crxHeader.pubKeyLength + crxHeader.signatureLength);
      }).then(function (data) {
        crxHeader.publicKey = data.slice(0, crxHeader.pubKeyLength);
        crxHeader.signature = data.slice(crxHeader.pubKeyLength);
        crxHeader.size = 16 + crxHeader.pubKeyLength + crxHeader.signatureLength;
        return crxHeader;
      });
    }
  });
}

// Zip64 File Format Notes: https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT
function getZip64CentralDirectory(source, zip64CDL) {
  var d64loc = (0, _parseBuffer["default"])(zip64CDL, [['signature', 4], ['diskNumber', 4], ['offsetToStartOfCentralDirectory', 8], ['numberOfDisks', 4]]);
  if (d64loc.signature != 0x07064b50) {
    throw new Error('invalid zip64 end of central dir locator signature (0x07064b50): 0x' + d64loc.signature.toString(16));
  }
  var dir64 = new _PullStream["default"]();
  source.stream(d64loc.offsetToStartOfCentralDirectory).pipe(dir64);
  return dir64.pull(56);
}

// Zip64 File Format Notes: https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT
function parseZip64DirRecord(dir64record) {
  var vars = (0, _parseBuffer["default"])(dir64record, [['signature', 4], ['sizeOfCentralDirectory', 8], ['version', 2], ['versionsNeededToExtract', 2], ['diskNumber', 4], ['diskStart', 4], ['numberOfRecordsOnDisk', 8], ['numberOfRecords', 8], ['sizeOfCentralDirectory', 8], ['offsetToStartOfCentralDirectory', 8]]);
  if (vars.signature != 0x06064b50) {
    throw new Error('invalid zip64 end of central dir locator signature (0x06064b50): 0x0' + vars.signature.toString(16));
  }
  return vars;
}
function centralDirectory(source, options) {
  var endDir = new _PullStream["default"]();
  var records = new _PullStream["default"]();
  var tailSize = options && options.tailSize || 80;
  var sourceSize, crxHeader, startOffset, vars;
  if (options && options.crx) crxHeader = getCrxHeader(source);
  return source.size().then(function (size) {
    sourceSize = size;
    source.stream(Math.max(0, size - tailSize)).on('error', function (error) {
      endDir.emit('error', error);
    }).pipe(endDir);
    return endDir.pull(signature);
  }).then(function () {
    return Promise.all([endDir.pull(22), crxHeader]);
  }).then(function (_ref) {
    var _ref2 = _slicedToArray(_ref, 2),
      directory = _ref2[0],
      crxHeader = _ref2[1];
    var data = directory;
    startOffset = crxHeader && crxHeader.size || 0;
    vars = (0, _parseBuffer["default"])(data, [['signature', 4], ['diskNumber', 2], ['diskStart', 2], ['numberOfRecordsOnDisk', 2], ['numberOfRecords', 2], ['sizeOfCentralDirectory', 4], ['offsetToStartOfCentralDirectory', 4], ['commentLength', 2]]);

    // Is this zip file using zip64 format? Use same check as Go:
    // https://github.com/golang/go/blob/master/src/archive/zip/reader.go#L503
    // For zip64 files, need to find zip64 central directory locator header to extract
    // relative offset for zip64 central directory record.
    if (vars.diskNumber == 0xffff || vars.numberOfRecords == 0xffff || vars.offsetToStartOfCentralDirectory == 0xffffffff) {
      // Offset to zip64 CDL is 20 bytes before normal CDR
      var zip64CDLSize = 20;
      var zip64CDLOffset = sourceSize - (tailSize - endDir.match + zip64CDLSize);
      var zip64CDLStream = new _PullStream["default"]();
      source.stream(zip64CDLOffset).pipe(zip64CDLStream);
      return zip64CDLStream.pull(zip64CDLSize).then(function (d) {
        return getZip64CentralDirectory(source, d);
      }).then(function (dir64record) {
        vars = parseZip64DirRecord(dir64record);
      });
    } else {
      vars.offsetToStartOfCentralDirectory += startOffset;
    }
  }).then(function () {
    if (vars.commentLength) return endDir.pull(vars.commentLength).then(function (comment) {
      vars.comment = comment.toString('utf8');
    });
  }).then(/*#__PURE__*/_asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee3() {
    var i;
    return _regenerator().w(function (_context3) {
      while (1) switch (_context3.n) {
        case 0:
          source.stream(vars.offsetToStartOfCentralDirectory).pipe(records);
          vars.extract = /*#__PURE__*/function () {
            var _ref4 = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee2(opts) {
              var files;
              return _regenerator().w(function (_context2) {
                while (1) switch (_context2.n) {
                  case 0:
                    if (!(!opts || !opts.path)) {
                      _context2.n = 1;
                      break;
                    }
                    throw new Error('PATH_MISSING');
                  case 1:
                    // make sure path is normalized before using it
                    opts.path = _path["default"].resolve(_path["default"].normalize(opts.path));
                    files = vars.files;
                    _context2.n = 2;
                    return (0, _mapPromises["default"])(files, /*#__PURE__*/function () {
                      var _ref5 = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee(entry) {
                        var extractPath, writer;
                        return _regenerator().w(function (_context) {
                          while (1) switch (_context.n) {
                            case 0:
                              // to avoid zip slip (writing outside of the destination), we resolve
                              // the target path, and make sure it's nested in the intended
                              // destination, or not extract it otherwise.
                              extractPath = _path["default"].join(opts.path, entry.path);
                              if (!(extractPath.indexOf(opts.path) != 0)) {
                                _context.n = 1;
                                break;
                              }
                              return _context.a(2);
                            case 1:
                              if (!(entry.type == 'Directory')) {
                                _context.n = 3;
                                break;
                              }
                              _context.n = 2;
                              return (0, _ensureDir["default"])(extractPath);
                            case 2:
                              return _context.a(2);
                            case 3:
                              _context.n = 4;
                              return (0, _ensureDir["default"])(_path["default"].dirname(extractPath));
                            case 4:
                              writer = opts.getWriter ? opts.getWriter({
                                path: extractPath
                              }) : _fs["default"].createWriteStream(extractPath);
                              return _context.a(2, new Promise(function (resolve, reject) {
                                entry.stream(opts.password).on('error', reject).pipe(writer).on('close', resolve).on('error', reject);
                              }));
                          }
                        }, _callee);
                      }));
                      return function (_x2) {
                        return _ref5.apply(this, arguments);
                      };
                    }(), {
                      concurrency: opts.concurrency > 1 ? opts.concurrency : 1
                    });
                  case 2:
                    return _context2.a(2, _context2.v);
                }
              }, _callee2);
            }));
            return function (_x) {
              return _ref4.apply(this, arguments);
            };
          }();
          vars.files = [];
          i = 0;
        case 1:
          if (!(i < vars.numberOfRecords)) {
            _context3.n = 3;
            break;
          }
          _context3.n = 2;
          return records.pull(46).then(function (data) {
            var vars = (0, _parseBuffer["default"])(data, [['signature', 4], ['versionMadeBy', 2], ['versionsNeededToExtract', 2], ['flags', 2], ['compressionMethod', 2], ['lastModifiedTime', 2], ['lastModifiedDate', 2], ['crc32', 4], ['compressedSize', 4], ['uncompressedSize', 4], ['fileNameLength', 2], ['extraFieldLength', 2], ['fileCommentLength', 2], ['diskNumber', 2], ['internalFileAttributes', 2], ['externalFileAttributes', 4], ['offsetToLocalFileHeader', 4]]);
            vars.offsetToLocalFileHeader += startOffset;
            vars.lastModifiedDateTime = (0, _parseDateTime["default"])(vars.lastModifiedDate, vars.lastModifiedTime);
            return records.pull(vars.fileNameLength).then(function (fileNameBuffer) {
              vars.pathBuffer = fileNameBuffer;
              vars.path = fileNameBuffer.toString('utf8');
              vars.isUnicode = (vars.flags & 0x800) != 0;
              return records.pull(vars.extraFieldLength);
            }).then(function (extraField) {
              vars.extra = (0, _parseExtraField["default"])(extraField, vars);
              return records.pull(vars.fileCommentLength);
            }).then(function (comment) {
              vars.comment = comment;
              vars.type = vars.uncompressedSize === 0 && /[/\\]$/.test(vars.path) ? 'Directory' : 'File';
              var padding = options && options.padding || 1000;
              vars.stream = function (_password) {
                var totalSize = 30 + padding // add an extra buffer
                + (vars.extraFieldLength || 0) + (vars.fileNameLength || 0) + vars.compressedSize;
                return (0, _unzip["default"])(source, vars.offsetToLocalFileHeader, _password, vars, totalSize);
              };
              vars.buffer = function (_password) {
                return (0, _BufferStream["default"])(vars.stream(_password));
              };
              return vars;
            });
          });
        case 2:
          vars.files[i] = _context3.v;
          i++;
          _context3.n = 1;
          break;
        case 3:
          return _context3.a(2, vars);
      }
    }, _callee3);
  })));
}
;
