// unzipper-esm/dist/Open/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _gracefulFs = _interopRequireDefault(require("graceful-fs"));
var _stream = require("stream");
var _directory = _interopRequireDefault(require("./directory.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; } // 'node:stream'
var _default = exports["default"] = {
  buffer: function buffer(_buffer, options) {
    var source = {
      stream: function stream(offset, length) {
        var stream = new _stream.PassThrough();
        var end = length ? offset + length : undefined;
        stream.end(_buffer.slice(offset, end));
        return stream;
      },
      size: function size() {
        return Promise.resolve(_buffer.length);
      }
    };
    return (0, _directory["default"])(source, options);
  },
  file: function file(filename, options) {
    var source = {
      stream: function stream(start, length) {
        var end = length ? start + length : undefined;
        return _gracefulFs["default"].createReadStream(filename, {
          start: start,
          end: end
        });
      },
      size: function size() {
        return new Promise(function (resolve, reject) {
          _gracefulFs["default"].stat(filename, function (err, d) {
            if (err) reject(err);else resolve(d.size);
          });
        });
      }
    };
    return (0, _directory["default"])(source, options);
  },
  url: function url(request, params, options) {
    if (typeof params === 'string') params = {
      url: params
    };
    if (!params.url) throw 'URL missing';
    params.headers = params.headers || {};
    var source = {
      stream: function stream(offset, length) {
        var options = Object.create(params);
        var end = length ? offset + length : '';
        options.headers = Object.create(params.headers);
        options.headers.range = 'bytes=' + offset + '-' + end;
        return request(options);
      },
      size: function size() {
        return new Promise(function (resolve, reject) {
          var req = request(params);
          req.on('response', function (d) {
            req.abort();
            if (!d.headers['content-length']) reject(new Error('Missing content length header'));else resolve(d.headers['content-length']);
          }).on('error', reject);
        });
      }
    };
    return (0, _directory["default"])(source, options);
  },
  s3: function s3(client, params, options) {
    var source = {
      size: function size() {
        return new Promise(function (resolve, reject) {
          client.headObject(params, function (err, d) {
            if (err) reject(err);else resolve(d.ContentLength);
          });
        });
      },
      stream: function stream(offset, length) {
        var d = {};
        for (var key in params) d[key] = params[key];
        var end = length ? offset + length : '';
        d.Range = 'bytes=' + offset + '-' + end;
        return client.getObject(d).createReadStream();
      }
    };
    return (0, _directory["default"])(source, options);
  },
  s3_v3: function s3_v3(client, params, options) {
    // `GetObjectCommand` and `HeadObjectCommand` from "@aws-sdk/client-s3" package
    // are not included in the distribution by default. The authors said:
    //
    // "To keep node-unzipper super small, a decision was made to not include optional
    //  third party sdks as a part of the library itself. unzipper has plenty of users
    //  that do not require the s3 features and it would be very inefficient to force them to do so".
    //
    var _global = global,
      GetObjectCommand = _global.GetObjectCommand,
      HeadObjectCommand = _global.HeadObjectCommand;
    if (!GetObjectCommand || !HeadObjectCommand) {
      // throw new Error('AWS S3 v3 support should be exported separately. See: https://github.com/ZJONSSON/node-unzipper/issues/330');
      throw new Error('You must set `global.GetObjectCommand` and `global.HeadObjectCommand` variables first');
    }
    var source = {
      size: function () {
        var _size = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee() {
          var head;
          return _regenerator().w(function (_context) {
            while (1) switch (_context.n) {
              case 0:
                _context.n = 1;
                return client.send(new HeadObjectCommand({
                  Bucket: params.Bucket,
                  Key: params.Key
                }));
              case 1:
                head = _context.v;
                if (head.ContentLength) {
                  _context.n = 2;
                  break;
                }
                return _context.a(2, 0);
              case 2:
                return _context.a(2, head.ContentLength);
            }
          }, _callee);
        }));
        function size() {
          return _size.apply(this, arguments);
        }
        return size;
      }(),
      stream: function stream(offset, length) {
        var stream = new _stream.PassThrough();
        var end = length ? offset + length : "";
        client.send(new GetObjectCommand({
          Bucket: params.Bucket,
          Key: params.Key,
          Range: "bytes=".concat(offset, "-").concat(end)
        })).then(function (response) {
          response.Body.pipe(stream);
        })["catch"](function (error) {
          stream.emit("error", error);
        });
        return stream;
      }
    };
    return (0, _directory["default"])(source, options);
  },
  custom: function custom(source, options) {
    return (0, _directory["default"])(source, options);
  }
};
