// unzipper-esm/dist/Decrypt.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _nodeInt = _interopRequireDefault(require("node-int64"));
var _stream = require("stream");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
// "node:stream"
var Decrypt = exports["default"] = /*#__PURE__*/function () {
  function Decrypt() {
    _classCallCheck(this, Decrypt);
    this.key0 = Buffer.allocUnsafe(4);
    this.key1 = Buffer.allocUnsafe(4);
    this.key2 = Buffer.allocUnsafe(4);
    this.key0.writeUInt32BE(0x12345678, 0);
    this.key1.writeUInt32BE(0x23456789, 0);
    this.key2.writeUInt32BE(0x34567890, 0);
  }
  return _createClass(Decrypt, [{
    key: "update",
    value: function update(h) {
      this.key0.writeUInt32BE(crc(h, this.key0));
      this.key1.writeUInt32BE((this.key0.readUInt32BE() & 0xff & 0xFFFFFFFF) + this.key1.readUInt32BE() >>> 0);
      var x = new _nodeInt["default"](multiply(this.key1.readUInt32BE(), 134775813) + 1 & 0xFFFFFFFF);
      var b = Buffer.alloc(8);
      x.copy(b, 0);
      b.copy(this.key1, 0, 4, 8);
      this.key2.writeUInt32BE(crc((this.key1.readUInt32BE() >> 24 & 0xff) >>> 0, this.key2));
    }
  }, {
    key: "decryptByte",
    value: function decryptByte(c) {
      var k = (this.key2.readUInt32BE() | 2) >>> 0;
      c = c ^ multiply(k, k ^ 1 >>> 0) >> 8 & 0xff;
      this.update(c);
      return c;
    }
  }, {
    key: "stream",
    value: function stream() {
      var self = this;
      return new _stream.Transform({
        transform: function transform(d, e, cb) {
          for (var i = 0; i < d.length; i++) {
            d[i] = self.decryptByte(d[i]);
          }
          this.push(d);
          cb();
        }
      });
    }
  }]);
}();
var table;
function generateTable() {
  var poly = 0xEDB88320;
  var c, n, k;
  table = [];
  for (n = 0; n < 256; n++) {
    c = n;
    for (k = 0; k < 8; k++) c = c & 1 ? poly ^ c >>> 1 : c = c >>> 1;
    table[n] = c >>> 0;
  }
}
function crc(ch, crc) {
  if (!table) generateTable();
  if (ch.charCodeAt) ch = ch.charCodeAt(0);
  var l = crc.readUInt32BE() >> 8 & 0xffffff;
  var r = table[(crc.readUInt32BE() ^ ch >>> 0) & 0xff];
  return (l ^ r) >>> 0;
}
function multiply(a, b) {
  var ah = a >> 16 & 0xffff;
  var al = a & 0xffff;
  var bh = b >> 16 & 0xffff;
  var bl = b & 0xffff;
  var high = ah * bl + al * bh & 0xffff;
  return (high << 16 >>> 0) + al * bl;
}
