// unzipper-esm/dist/extract.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Extract;
var _fs = _interopRequireDefault(require("fs"));
var _path = _interopRequireDefault(require("path"));
var _stream = require("stream");
var _DuplexStream = _interopRequireDefault(require("./DuplexStream.js"));
var _parse = _interopRequireDefault(require("./parse.js"));
var _ensureDir = _interopRequireDefault(require("./ensureDir.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; } // 'node:fs'
// 'node:path'
// 'node:stream'
function Extract(opts) {
  // make sure path is normalized before using it
  opts.path = _path["default"].resolve(_path["default"].normalize(opts.path));
  var parser = (0, _parse["default"])(opts);
  var outStream = new _stream.Writable({
    objectMode: true,
    write: function () {
      var _write = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee(entry, encoding, cb) {
        var extractPath, writer;
        return _regenerator().w(function (_context) {
          while (1) switch (_context.n) {
            case 0:
              // to avoid zip slip (writing outside of the destination), we resolve
              // the target path, and make sure it's nested in the intended
              // destination, or not extract it otherwise.
              // NOTE: Need to normalize to forward slashes for UNIX OS's to properly
              // ignore the zip slipped file entirely
              extractPath = _path["default"].join(opts.path, entry.path.replace(/\\/g, '/'));
              if (!(extractPath.indexOf(opts.path) != 0)) {
                _context.n = 1;
                break;
              }
              return _context.a(2, cb());
            case 1:
              if (!(entry.type == 'Directory')) {
                _context.n = 3;
                break;
              }
              _context.n = 2;
              return (0, _ensureDir["default"])(extractPath);
            case 2:
              return _context.a(2, cb());
            case 3:
              _context.n = 4;
              return (0, _ensureDir["default"])(_path["default"].dirname(extractPath));
            case 4:
              writer = opts.getWriter ? opts.getWriter({
                path: extractPath
              }) : _fs["default"].createWriteStream(extractPath);
              entry.pipe(writer).on('error', cb).on('close', cb);
            case 5:
              return _context.a(2);
          }
        }, _callee);
      }));
      function write(_x, _x2, _x3) {
        return _write.apply(this, arguments);
      }
      return write;
    }()
  });

  // Create a combined stream
  var extract = new _DuplexStream["default"](parser, outStream);
  parser.once('crx-header', function (crxHeader) {
    extract.crxHeader = crxHeader;
  });
  parser.pipe(outStream).on('finish', function () {
    extract.emit('close');
  });
  extract.promise = function () {
    return new Promise(function (resolve, reject) {
      extract.on('close', resolve);
      extract.on('error', reject);
    });
  };
  return extract;
}
