// unzipper-esm/dist/parse.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Parse;
var _zlib = _interopRequireDefault(require("zlib"));
var _stream = require("stream");
var _PullStream2 = _interopRequireDefault(require("./PullStream.js"));
var _NoopStream = _interopRequireDefault(require("./NoopStream.js"));
var _BufferStream = _interopRequireDefault(require("./BufferStream.js"));
var _parseExtraField = _interopRequireDefault(require("./parseExtraField.js"));
var _parseDateTime = _interopRequireDefault(require("./parseDateTime.js"));
var _parseBuffer = _interopRequireDefault(require("./parseBuffer.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _possibleConstructorReturn(t, e) { if (e && ("object" == _typeof(e) || "function" == typeof e)) return e; if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined"); return _assertThisInitialized(t); }
function _assertThisInitialized(e) { if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); return e; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _getPrototypeOf(t) { return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) { return t.__proto__ || Object.getPrototypeOf(t); }, _getPrototypeOf(t); }
function _inherits(t, e) { if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function"); t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && _setPrototypeOf(t, e); }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); } // 'node:zlib'
// 'node:stream'
var endDirectorySignature = Buffer.alloc(4);
endDirectorySignature.writeUInt32LE(0x06054b50, 0);
function Parse(opts) {
  return new Parser(opts);
}
var Parser = /*#__PURE__*/function (_PullStream) {
  function Parser() {
    var _this;
    var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
      verbose: false
    };
    _classCallCheck(this, Parser);
    _this = _callSuper(this, Parser, [opts]);
    _this._opts = opts;
    var self = _this;
    self.on('finish', function () {
      self.emit('end');
      self.emit('close');
    });
    self._readRecord()["catch"](function (e) {
      if (!self.__emittedError || self.__emittedError !== e) self.emit('error', e);
    });
    return _this;
  }
  _inherits(Parser, _PullStream);
  return _createClass(Parser, [{
    key: "_readRecord",
    value: function _readRecord() {
      var self = this;
      return self.pull(4).then(function (data) {
        if (data.length === 0) return;
        var signature = data.readUInt32LE(0);
        if (signature === 0x34327243) {
          return self._readCrxHeader();
        }
        if (signature === 0x04034b50) {
          return self._readFile();
        } else if (signature === 0x02014b50) {
          self.reachedCD = true;
          return self._readCentralDirectoryFileHeader();
        } else if (signature === 0x06054b50) {
          return self._readEndOfCentralDirectoryRecord();
        } else if (self.reachedCD) {
          // _readEndOfCentralDirectoryRecord expects the EOCD
          // signature to be consumed so set includeEof=true
          var includeEof = true;
          return self.pull(endDirectorySignature, includeEof).then(function () {
            return self._readEndOfCentralDirectoryRecord();
          });
        } else self.emit('error', new Error('invalid signature: 0x' + signature.toString(16)));
      }).then(function (loop) {
        if (loop) {
          return self._readRecord();
        }
      });
    }
  }, {
    key: "_readCrxHeader",
    value: function _readCrxHeader() {
      var self = this;
      return self.pull(12).then(function (data) {
        self.crxHeader = (0, _parseBuffer["default"])(data, [['version', 4], ['pubKeyLength', 4], ['signatureLength', 4]]);
        return self.pull(self.crxHeader.pubKeyLength + self.crxHeader.signatureLength);
      }).then(function (data) {
        self.crxHeader.publicKey = data.slice(0, self.crxHeader.pubKeyLength);
        self.crxHeader.signature = data.slice(self.crxHeader.pubKeyLength);
        self.emit('crx-header', self.crxHeader);
        return true;
      });
    }
  }, {
    key: "_readFile",
    value: function _readFile() {
      var self = this;
      return self.pull(26).then(function (data) {
        var vars = (0, _parseBuffer["default"])(data, [['versionsNeededToExtract', 2], ['flags', 2], ['compressionMethod', 2], ['lastModifiedTime', 2], ['lastModifiedDate', 2], ['crc32', 4], ['compressedSize', 4], ['uncompressedSize', 4], ['fileNameLength', 2], ['extraFieldLength', 2]]);
        vars.lastModifiedDateTime = (0, _parseDateTime["default"])(vars.lastModifiedDate, vars.lastModifiedTime);
        if (self.crxHeader) vars.crxHeader = self.crxHeader;
        return self.pull(vars.fileNameLength).then(function (fileNameBuffer) {
          var fileName = fileNameBuffer.toString('utf8');
          var entry = new _stream.PassThrough();
          var __autodraining = false;
          entry.autodrain = function () {
            __autodraining = true;
            var draining = entry.pipe(new _NoopStream["default"]());
            draining.promise = function () {
              return new Promise(function (resolve, reject) {
                draining.on('finish', resolve);
                draining.on('error', reject);
              });
            };
            return draining;
          };
          entry.buffer = function () {
            return (0, _BufferStream["default"])(entry);
          };
          entry.path = fileName;
          entry.props = {};
          entry.props.path = fileName;
          entry.props.pathBuffer = fileNameBuffer;
          entry.props.flags = {
            "isUnicode": (vars.flags & 0x800) != 0
          };
          entry.type = vars.uncompressedSize === 0 && /[/\\]$/.test(fileName) ? 'Directory' : 'File';
          if (self._opts.verbose) {
            if (entry.type === 'Directory') {
              console.log('   creating:', fileName);
            } else if (entry.type === 'File') {
              if (vars.compressionMethod === 0) {
                console.log(' extracting:', fileName);
              } else {
                console.log('  inflating:', fileName);
              }
            }
          }
          return self.pull(vars.extraFieldLength).then(function (extraField) {
            var extra = (0, _parseExtraField["default"])(extraField, vars);
            entry.vars = vars;
            entry.extra = extra;
            if (self._opts.forceStream) {
              self.push(entry);
            } else {
              self.emit('entry', entry);
              if (self._readableState.pipesCount || self._readableState.pipes && self._readableState.pipes.length) self.push(entry);
            }
            if (self._opts.verbose) console.log({
              filename: fileName,
              vars: vars,
              extra: extra
            });
            var fileSizeKnown = !(vars.flags & 0x08) || vars.compressedSize > 0;
            var eof;
            entry.__autodraining = __autodraining; // expose __autodraining for test purposes
            var inflater = vars.compressionMethod && !__autodraining ? _zlib["default"].createInflateRaw() : (0, _stream.PassThrough)();
            if (fileSizeKnown) {
              entry.size = vars.uncompressedSize;
              eof = vars.compressedSize;
            } else {
              eof = Buffer.alloc(4);
              eof.writeUInt32LE(0x08074b50, 0);
            }
            return new Promise(function (resolve, reject) {
              (0, _stream.pipeline)(self.stream(eof), inflater, entry, function (err) {
                if (err) {
                  return reject(err);
                }
                return fileSizeKnown ? resolve(fileSizeKnown) : self._processDataDescriptor(entry).then(resolve)["catch"](reject);
              });
            });
          });
        });
      });
    }
  }, {
    key: "_processDataDescriptor",
    value: function _processDataDescriptor(entry) {
      var self = this;
      return self.pull(16).then(function (data) {
        var vars = (0, _parseBuffer["default"])(data, [['dataDescriptorSignature', 4], ['crc32', 4], ['compressedSize', 4], ['uncompressedSize', 4]]);
        entry.size = vars.uncompressedSize;
        return true;
      });
    }
  }, {
    key: "_readCentralDirectoryFileHeader",
    value: function _readCentralDirectoryFileHeader() {
      var self = this;
      return self.pull(42).then(function (data) {
        var vars = (0, _parseBuffer["default"])(data, [['versionMadeBy', 2], ['versionsNeededToExtract', 2], ['flags', 2], ['compressionMethod', 2], ['lastModifiedTime', 2], ['lastModifiedDate', 2], ['crc32', 4], ['compressedSize', 4], ['uncompressedSize', 4], ['fileNameLength', 2], ['extraFieldLength', 2], ['fileCommentLength', 2], ['diskNumber', 2], ['internalFileAttributes', 2], ['externalFileAttributes', 4], ['offsetToLocalFileHeader', 4]]);
        return self.pull(vars.fileNameLength).then(function (fileName) {
          vars.fileName = fileName.toString('utf8');
          return self.pull(vars.extraFieldLength);
        }).then(function () {
          return self.pull(vars.fileCommentLength);
        }).then(function () {
          return true;
        });
      });
    }
  }, {
    key: "_readEndOfCentralDirectoryRecord",
    value: function _readEndOfCentralDirectoryRecord() {
      var self = this;
      return self.pull(18).then(function (data) {
        var vars = (0, _parseBuffer["default"])(data, [['diskNumber', 2], ['diskStart', 2], ['numberOfRecordsOnDisk', 2], ['numberOfRecords', 2], ['sizeOfCentralDirectory', 4], ['offsetToStartOfCentralDirectory', 4], ['commentLength', 2]]);
        return self.pull(vars.commentLength).then(function () {
          self.end();
          self.push(null);
        });
      });
    }
  }, {
    key: "promise",
    value: function promise() {
      var self = this;
      return new Promise(function (resolve, reject) {
        self.on('finish', resolve);
        self.on('error', reject);
      });
    }
  }]);
}(_PullStream2["default"]);
